### PR TITLE
Update doc for sparse related APIs

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
- update NDArrayIter API doc with CSRNDArray inputs
- update document for kv.init, kv.push 
- add test for #7676
preview: http://ec2-54-187-32-207.us-west-2.compute.amazonaws.com/api/python/optimization.html#mxnet.optimizer.SGD

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
@anirudh2290 @rahul003 @reminisce

**Comment by eric-haibin-lin** [comment]
Also updated doc for optimizers. Please help review.

**Comment by eric-haibin-lin** [comment]
@cjolivier01

File: python/mxnet/kvstore.py [hunk 1]
```diff
@@ -120,8 +120,15 @@ def init(self, key, value):
...
 
         >>> # init a list of key-value pairs
-        >>> keys = [5, 7, 9]
+        >>> keys = ['5', '7', '9']
         >>> kv.init(keys, [mx.nd.ones(shape)]*len(keys))
+
+        >>> # init a row_sparse value
+        >>> kv.init('4', mx.nd.ones(shape).tostype('row_sparse'))
+        >>> b = mx.nd.sparse.zeros('row_sparse', shape)
+        >>> kv.row_sparse_pull('4', row_ids=mx.nd.array([0, 1]), out=b)
```
Comment (anirudh2290) [inline python/mxnet/kvstore.py]: Without giving the dtype I get the error: "Unknown type enum 0". Would it make sense to pull just one of the rows out of the two as the example ?

File: python/mxnet/kvstore.py [hunk 2]
```diff
@@ -187,6 +195,15 @@ def push(self, key, value, priority=0):
         >>> print b[1][1].asnumpy()
         [[ 4.  4.  4.]
         [ 4.  4.  4.]]
+
+        >>> # push a row_sparse value
+        >>> b = mx.nd.sparse.zeros('row_sparse', shape)
+        >>> kv.init('10', mx.nd.sparse.zeros('row_sparse', shape))
+        >>> kv.push('10', mx.nd.ones(shape).tostype('row_sparse'))
+        >>> # pull out the value
+        >>> kv.row_sparse_pull('10', row_ids=mx.nd.array([0, 1]), out=b)
```
Comment (anirudh2290) [inline python/mxnet/kvstore.py]: same error as above

File: python/mxnet/io.py [hunk 3]
```diff
@@ -603,6 +617,11 @@ def __init__(self, data, label=None, batch_size=1, shuffle=False,
 
         self.data = _init_data(data, allow_empty=False, default_name=data_name)
         self.label = _init_data(label, allow_empty=True, default_name=label_name)
+        if isinstance(data, CSRNDArray) or isinstance(label, CSRNDArray):
+            assert(last_batch_handle == 'discard'), \
+                  "`NDArrayIter` only supports ``CSRNDArray`` with `shuffle` set to `False`"
```
[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply python/mxnet/io.py]: Good catch. I'll correct it

File: python/mxnet/kvstore.py (hunk 1 above)
Comment (eric-haibin-lin) [reply python/mxnet/kvstore.py]: `Unknown type enum 0 ` -> this is fixed in this PR if you checkout my code. (see changes in kvstore_local.h)
In this example, I'm showing all the rows initialized so I gave all row_ids. In the example of `row_sparse_pull`, there's an example of pulling a subset of the rows.

File: python/mxnet/kvstore.py (hunk 2 above)
Comment (eric-haibin-lin) [reply python/mxnet/kvstore.py]: see https://github.com/apache/incubator-mxnet/pull/7688/files/72440bfca52c93c79d25d8fcacc457cdbc7c872f#diff-5df45b6fc6d476aa1e0943f656c88354R219

File: python/mxnet/io.py [hunk 4]
```diff
@@ -513,7 +514,8 @@ def _init_data(data, allow_empty, default_name):
     return list(data.items())
 
 class NDArrayIter(DataIter):
-    """Returns an iterator for ``mx.nd.NDArray``, ``numpy.ndarray`` or ``h5py.Dataset``.
+    """Returns an iterator for ``mx.nd.NDArray``, ``numpy.ndarray``, ``h5py.Dataset``
+    or ``mx.nd.sparse.CSRNDArray``.
```
[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply python/mxnet/io.py]: No we don't have any

File: python/mxnet/io.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply python/mxnet/io.py]: Updated now.

File: python/mxnet/kvstore.py [hunk 5]
```diff
@@ -266,8 +283,8 @@ def pull(self, key, out=None, priority=0):
                 self.handle, mx_uint(len(ckeys)), ckeys, cvals, ctypes.c_int(priority)))
 
     def row_sparse_pull(self, key, out=None, priority=0, row_ids=None):
-        """ Pulls a single row_sparse value or a sequence of row_sparse values from the store
-         with specified row_ids.
+        """ Pulls a single RowSparseNDArray value or a sequence of RowSparseNDArray values \
```
Comment (reminisce) [inline python/mxnet/kvstore.py]: Why need "\" at the end of the line?

Comment (eric-haibin-lin) [reply python/mxnet/kvstore.py]: causes display problem at API doc. Otherwise, it doesn't treat the second line as part of the first line

**Pull request merged**